src/mame/machine/k573fpga.cpp

```diff
@@ -0,0 +1,592 @@
+            // The game is still requesting position updates after the song has ended.
+            // This happens in some games like DDR 4th Mix Plus where it uses
+            // the position in song for actual game engine timings.
+            // The counter will properly end when the game signals to the FPGA to stop playback.
+            last_position_update += position_diff;
+        } else {
+            position_diff = m_samples->get_position(0) - last_position_update;
+            last_position_update = m_samples->get_position(0);
+        }
+
+        last_position_update = m_samples->get_position(0);
+
+        return last_position_update;
+    }
+
+    return 0;
+}
+
+uint16_t k573fpga_device::i2c_read()
+{
+    int scl = mas3507d->i2c_scl_r() << 13;
+    int sda = mas3507d->i2c_sda_r() << 12;
+
+    return scl | sda;
+}
+
+void k573fpga_device::i2c_write(uint16_t data)
+{
+    mas3507d->i2c_scl_w(data & 0x2000);
+    mas3507d->i2c_sda_w(data & 0x1000);
+}
+
+uint16_t k573fpga_device::get_mpeg_ctrl()
+{
+    if (mpeg_ctrl_flag == 0xe000 || (m_samples->get_position(0) > 0 && mp3_decrypt_mode)) {
+        // This has been tested with real hardware, but this flag is always held 0x1000 when the audio is being played
+        return mpeg_ctrl_flag | 0x1000;
+    }
+
+    return mpeg_ctrl_flag;
+}
+
+void k573fpga_device::set_mpeg_ctrl(uint16_t data)
+{
+    logerror("FPGA MPEG control %c%c%c | %08x %08x\n",
+                data & 0x8000 ? '#' : '.',
+                data & 0x4000 ? '#' : '.',
+                data & 0x2000 ? '#' : '.',
+                mp3_start_adr, mp3_end_adr);
+
+    mpeg_ctrl_flag = data;
+
+    if (data == 0xa000) {
+        // Stop playback
+        m_samples->stop(0);
+        m_samples->stop(1);
+    }
+
+    if ((data & 0xa000) == 0xa000) {
+        // Reset playback state for start and stop events
+        mp3_last_frame = 0;
+        mp3_next_sync = 0;
+        mp3_last_adr = mp3_start_adr;
+        mp3_last_decrypt_adr = mp3_start_adr;
+        last_copied_samples = 0;
+        last_buffer_size_channel_l = 0;
+        last_buffer_size_channel_r = 0;
+
+        last_position_update = 0;
+        position_diff = 0;
+        decrypt_finished = false;
+
+        if (channel_l_pcm) {
+            free(channel_l_pcm);
+            channel_l_pcm = nullptr;
+        }
+
+        if (channel_r_pcm) {
+            free(channel_r_pcm);
+            channel_r_pcm = nullptr;
+        }
+    }
+
+    if ((data & 0xe000) == 0xe000) {
+        mp3_decrypt_mode = true;
+    } else {
+        mp3_decrypt_mode = false;
+    }
+}
+
+int32_t k573fpga_device::find_enc_key()
+{
+    if (k573enc_lookup.find(crypto_key1) == k573enc_lookup.end()) {
+        return -1;
+    }
+
+    return k573enc_lookup[crypto_key1];
+}
+
+
+inline uint8_t IsBitSet(int value, int n)
+{
+    return (value >> n) & 1;
+}
+
+inline uint16_t BitSwap16(int v, int b15, int b14, int b13, int b12, int b11, int b10, int b9, int b8,
```

Comment:
We have templates in MAME for that, please use them.  I'm referring to BIT, BITSWAP16, etc.

Reply:
Thank you. I found the functions in the utils class and made the appropriate changes and confirmed everything still works. :+1: